### PR TITLE
Feature/typescript project prop context - DO NOT MERGE

### PR DESCRIPTION
**Pull request by norbertschuler**
Extended version of #1653 with some Typescript changes using lowercase branch name.

### CODE REVIEW

File: src/features/common/Layout/ProjectPropsContext.tsx
Comment (mohitb35) [inline src/features/common/Layout/ProjectPropsContext.tsx:187]: It would be better to avoid creating this `defaultState` altogether. `ProjectPropsProvider` initializes the context using local state anyway. That will also avoid the following warning that is seen in the console:

```Argument of type '{ project: {} | null; setProject: () => void; projects: never[]; setProjects: () => void; plantLocations: never[]; setPlantLocations: () => void; showSingleProject: boolean; setShowSingleProject: () => void; ... 66 more ...; setIsMobile: () => void; }' is not assignable to parameter of type 'ProjectContextInterface'.
  Types of property 'project' are incompatible.
    Type '{} | null' is not assignable to type 'Project | null'.
      Type '{}' is missing the following properties from type 'Project': id, _scope, allowDonations, classification, and 44 more. ```

Simply initialize the context as null while creating it.

Comment (mohitb35) [reply src/features/common/Layout/ProjectPropsContext.tsx:187]: It would be better to avoid creating this `defaultState` altogether. `ProjectPropsProvider` initializes the context using local state anyway. That will also avoid the following warning that is seen in the console on line 190:

```Argument of type '{ project: {} | null; setProject: () => void; projects: never[]; setProjects: () => void; plantLocations: never[]; setPlantLocations: () => void; showSingleProject: boolean; setShowSingleProject: () => void; ... 66 more ...; setIsMobile: () => void; }' is not assignable to parameter of type 'ProjectContextInterface'.
  Types of property 'project' are incompatible.
    Type '{} | null' is not assignable to type 'Project | null'.
      Type '{}' is missing the following properties from type 'Project': id, _scope, allowDonations, classification, and 44 more. ```

Simply initialize the context as null while creating it.

File: src/features/common/Layout/ProjectPropsContext.tsx
Comment (mohitb35) [inline src/features/common/Layout/ProjectPropsContext.tsx:190]: It would be better to avoid creating a `defaultState` altogether. We are already setting up defaults using local state within `ProjectPropsProvider`. Simply initialize context as null while creating it.

This will also address the TS error on line 190.

```Argument of type '{ project: {} | null; setProject: () => void; projects: never[]; setProjects: () => void; plantLocations: never[]; setPlantLocations: () => void; showSingleProject: boolean; setShowSingleProject: () => void; ... 66 more ...; setIsMobile: () => void; }' is not assignable to parameter of type 'ProjectContextInterface'.
  Types of property 'project' are incompatible.
    Type '{} | null' is not assignable to type 'Project | null'.
      Type '{}' is missing the following properties from type 'Project': id, _scope, allowDonations, classification, and 44 more.```

Comment (mohitb35) [reply src/features/common/Layout/ProjectPropsContext.tsx:190]: Note: If you do this, you will need to handle the unlikely scenario that the context is `null`. This can be handled by creating a hook similar to how it is done for `BulkCodeProvider`.

```export const useBulkCode = (): BulkCodeContextInterface => {
  const context = useContext(BulkCodeContext);
  if (!context) {
    throw new Error('BulkCodeContext must be used within BulkCodeProvider');
  }
  return context;
};```

File: src/features/common/types/projectPropContext.d.ts
Comment (mohitb35) [inline src/features/common/types/projectPropContext.d.ts:1]: @sunilsabatp The types used here are not related to project props. I see types defined for Viewport, Map etc.

1. Please evaluate whether the type is actually needed. If it is for an external library, it may already be provided by that library.
2. Please separate the type definitions by their purpose into different files.

File: src/features/common/Layout/ProjectPropsContext.tsx
Comment (norbertschuler) [reply src/features/common/Layout/ProjectPropsContext.tsx:102]: @sunilsabatp can you take over this pull request again please.

File: src/features/common/Layout/ProjectPropsContext.tsx
Comment (prachigarg19) [inline src/features/common/Layout/ProjectPropsContext.tsx:187]: Remove the console log statements

File: src/features/common/Layout/ProjectPropsContext.tsx
Comment (prachigarg19) [inline src/features/common/Layout/ProjectPropsContext.tsx:165]: Argument of type mapStyle: is not assignable to parameter of type 'MapboxOptions | (() => MapboxOptions)'.
: The following warning is being displayed.

Comment (sunilsabatp) [reply src/features/common/Layout/ProjectPropsContext.tsx:167]: this issue has been resolved

File: src/features/common/types/projectPropContext.d.ts
Comment (sunilsabatp) [reply src/features/common/types/projectPropContext.d.ts:207]: now It is being use as  replacement for MapBoxOption  interface

**Review by prachigarg19 — APPROVED** [review]

**Comment by norbertschuler** [comment]
@sunilsabatp please tell me when I can merge this (or do it yourself if GitHub allows it)

**Comment by mariahosfeld** [comment]
Functionality seems to work correctly.

**Comment by norbertschuler** [comment]
@sunilsabatp can you merge this branch (as I am not owner, but originally it was you code)? Otherwise please ping me to let me merge it if it was approved.

**Comment by mohitb35** [comment]
@sunilsabatp @norbertschuler please do not merge this in. #1811 supersedes this PR.

**Pull request closed without merging**
